iroh/src/magicsock.rs

```diff
@@ -650,27 +652,86 @@ impl MagicSock {
         bufs: &mut [io::IoSliceMut<'_>],
         metas: &mut [quinn_udp::RecvMeta],
     ) -> Poll<io::Result<usize>> {
-        // FIXME: currently ipv4 load results in ipv6 traffic being ignored
         debug_assert_eq!(bufs.len(), metas.len(), "non matching bufs & metas");
         if self.is_closed() {
             return Poll::Pending;
         }
 
-        // order of polling is: UDPv4, UDPv6, relay
-        let (msgs, from_ipv4) = match self.pconn4.poll_recv(cx, bufs, metas)? {
-            Poll::Pending | Poll::Ready(0) => match &self.pconn6 {
-                Some(conn) => match conn.poll_recv(cx, bufs, metas)? {
-                    Poll::Pending | Poll::Ready(0) => {
-                        return self.poll_recv_relay(cx, bufs, metas);
+        // Three macros to help polling: they return if they get a result, execution
+        // continues if they were Pending and we need to poll others (or finally return
+        // Pending).
+        macro_rules! poll_ipv4 {
+            () => {
+                match self.pconn4.poll_recv(cx, bufs, metas)? {
+                    Poll::Pending | Poll::Ready(0) => {}
+                    Poll::Ready(n) => {
+                        self.process_udp_datagrams(true, &mut bufs[..n], &mut metas[..n]);
+                        return Poll::Ready(Ok(n));
                     }
-                    Poll::Ready(n) => (n, false),
-                },
-                None => {
-                    return self.poll_recv_relay(cx, bufs, metas);
                 }
-            },
-            Poll::Ready(n) => (n, true),
-        };
+            };
+        }
+        macro_rules! poll_ipv6 {
+            () => {
+                if let Some(ref pconn) = self.pconn6 {
+                    match pconn.poll_recv(cx, bufs, metas)? {
+                        Poll::Pending | Poll::Ready(0) => {}
+                        Poll::Ready(n) => {
+                            self.process_udp_datagrams(false, &mut bufs[..n], &mut metas[..n]);
+                            return Poll::Ready(Ok(n));
+                        }
+                    }
+                }
+            };
+        }
+        macro_rules! poll_relay {
+            () => {
+                match self.poll_recv_relay(cx, bufs, metas) {
+                    Poll::Pending => {}
+                    Poll::Ready(n) => return Poll::Ready(n),
+                }
+            };
+        }
+
+        let counter = self.poll_recv_counter.fetch_add(1, Ordering::Relaxed);
```

Comment:
is this an overflowing add? otherwise this could become an issue for long running sessions

Reply:
Yep, atomics wrap on overflow: https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU64.html#method.fetch_add

When I talked to @flub about this the last time, I wanted to go even a step further and be intentional about this: Make this an `AtomicU8`.

Reply:
So should it be U8?  Can we expect U8 to be more performant?

Choosing U8 means you get more bias though, since 16 isn't devisable by 3?

Reply:
if you want optimial performance, use `usize`, that is usually going to be most efficient

Reply:
> Choosing U8 means you get more bias though, since 16 isn't devisable by 3?

Yeah seems like you'll slightly bias towards 0: 255 % 3 = 0 and 0 % 3 = 0, too. All in all a slightly less than 1% bias though, I don't think it should matter for this case, right?

(If updated to `AtomicUsize` then the bias is much much smaller)

Reply:
Switched to AtomicUsize